lib/config/guiconfiguration.go

```diff
@@ -12,15 +12,28 @@ import (
 	"strings"
 )
 
+const (
+	AuthModeSimple = "simple"
+	AuthModeLDAP   = "ldap"
+)
+
+const (
+	LDAPTLSModeNonTSL   = "nontls"
+	LDAPTLSModeTSL      = "tls"
+	LDAPTLSModeStartTSL = "starttls"
+)
+
 type GUIConfiguration struct {
 	Enabled                   bool   `xml:"enabled,attr" json:"enabled" default:"true"`
 	RawAddress                string `xml:"address" json:"address" default:"127.0.0.1:8384"`
 	User                      string `xml:"user,omitempty" json:"user"`
 	Password                  string `xml:"password,omitempty" json:"password"`
 	AuthMode                  string `xml:"authMode,omitempty" json:"authMode"`
-	LdapServer                string `xml:"ldapServer,omitempty" json:"ldapServer"`
-	LdapPort                  int    `xml:"ldapPort,omitempty" json:"ldapPort"`
-	LdapBindDn                string `xml:"ldapBindDn,omitempty" json:"ldapBindDn"`
+	LDAPServer                string `xml:"LDAPServer,omitempty" json:"LDAPServer"`
+	LDAPPort                  int    `xml:"LDAPPort,omitempty" json:"LDAPPort"`
+	LDAPBindDn                string `xml:"LDAPBindDn,omitempty" json:"LDAPBindDn"`
+	LDAPTLSMode               string `xml:"LDAPTLSMode,omitempty" json:"LDAPTLSMode"`
+	LDAPInsecureSkipVerify    bool   `xml:"LDAPInsecureSkipVerify,attr" json:"LDAPInsecureSkipVerify" default:"false"`
```

Comment:
Again, discussion point before we make any changes.

If we are starting to add LDAP support to syncthing, perhaps we should have a separate LDAPConfiguration section in the config, and then GUIAuthMode == LDAP uses the LDAPConfiguration, and then once we add some other non-GUI feature that uses LDAP, we don't have to add yet another LDAP configuration to some other part of the config.


/cc @calmh

Reply:
Yeah, I agree with giving this thing its own config struct/section, and perhaps combining server+port into just an address, we don't typically separate the parts in other places of the config.

Reply:
I think you pick one or the other?

Reply:
Maybe. Or we fall back to trying static if LDAP is down? Can we distinguish "down" from "denied" based on the bind result?

Reply:
Technically Dial is up/down check and Bind is allowed/denied check and they are separate.

Reply:
Cool then it's easy to separate. So what's the desired outcome here? Assuming LDAP is configured,

- LDAP up, bind denied => denied
- LDAP down, static auth configured => follow static auth
- LDAP down, no static auth => denied (I guess? or allowed?)

Reply:
I personally prefer no fallback. If i set auth mode to be ldap and it does not work i would like know right away as an admin, while fallback would waste my time.
I would say:
1. Auth mode is set to ldap do only ldap
2. Auth mode is set to static do only static
3. Auth mode not set do static if login and password set (config backward compatibility for now)

Reply:
I am working on enums as suggested. Static is better, agree. Also extracting ldap config section. Will do another push soon.
1. Can you show me where the docs are?
2. Also I am still not sure how do you do deps. I had to do go get ./... but your TeamCity seems to not do that, so I had to include libs in vendor dir. Is this ok?

Reply:
Docs are at docs.syncthing.net, github.com/syncthing/docs. Dependencies need to be vendored, yes. We use gvt (https://github.com/FiloSottile/gvt). Essentially, `gvt fetch gopkg.in/whatever` and it'll add it to `vendor/` and update the manifest.